### DISCUSSION TITLE
remove skipped properties from serialized state

### DISCUSSION THREAD

**Discussion by sinisterstumble** [vuejs/pinia#2113] [Help and Questions]
### What problem is this solving

Pinia has the ability to skip the hydration of certain properties with `skipHydrate()`. However, there is currently no way to remove skipped properties from the serialized state.

### Proposed solution

Pinia could provide a function `filterSkipped` that would go over `pinia.state.value` and remove the skipped properties.

**Comment by posva** [comment]
What are you really trying to do? You shouldn't need to remove any properties from the serialized state. If that's the case, they probably shouldn't be in the state in the first place

**Reply by sinisterstumble** [reply]
It could be that I misunderstood something. I have a ref property in a pinia setup store that tracks a file loading state: `loaded | pending`. I use `skipHydrate` on that property since the client and the server must do their own loading. The property is used for UI interaction feedback on the client. Currently, the property is serialized into `pinia.state.value` as 'loaded', which is the server state, and it's incorrect for the client. The `skipHydrate` skips hydration of that property, and it ends up with the value 'pending' - the correct initial ref value for the client.

So everything works perfectly as-is. Removing the property from the serialized state would just free up a little bit of space.

**Reply by posva** [reply]
removing one or even a thousand numbers from memory won't make a difference in a Vue application, you don't need to worry about this, let JS handle the garbage collection 🙂

**Reply by sinisterstumble** [reply]
I meant the document size with the unnecessary properties in the serialized state.

**Comment by Csszabi98** [comment]
Hey, I have stumbled upon the need for such a functionality myself. My use case is the following:

1. Apollo queries are executed, and apollo takes care of caching the results, apollo state is restored on the client and the client uses the cache
2. State derived from the apollo query results is stored inside of pinia, and is used by the components to render
3. When I try to SSR my page, I end up with a duplicated state on the client that is quite large: `pinia.state` and `apollo.state`

The need arises to deduplicate this, as the `pinia` state essentially acts as "writableComputed" since it is derived from the apollo state. In fact, using and internal ref and exposing a computed derivate does solve the duplication, but it creates some overhead.

I have to define stores like this:
```js
const store  = defineStore('example', () => {
  const apolloDerivate = ref()

  function setApolloDerivate(newValue) {
    // Not so trivial state setting logic based on multiple refs
  }

  return {
    apolloDerivate: computed(() => apolloDerivate.value),
    setApolloDerivate
  }
})
```

Or use a utility function like:

```js
function nonSerializableRef() {
  const dataRef = ref()
 
  return computed({
    get() {
      return dataRef.value
    },
   set(newValue) {
     dataRef.value = newValue
   }
}
```

Both solutions are quite verbose and bring the overhead (and performance issues) of an extra computed property, when in fact what I would like to do is to just use something like `skipSerialize`.

The component rendering logic takes care of building up the store correctly during the hydration process using the instantly resolved Apollo query result (from the cache).

The components are architected this way to enable multiple data provider components upward the tree.

**Reply by posva** [reply]
Using a `computed` is as long as using a helper called `skipSerialize`. Note you could also use just a js getter or `toRaw()` to keep things reactive while skipping pinia's check for state (isRef or isReactive) [see playground](https://play.vuejs.org/#eNp9Uk1vwjAM/StWLgOtKpO4VTDtQxy2wzaxHXPpgmGBNImSFJCq/vc5CV+HDalSEr/n52e7HXu0tty2yCo28cJJG8BjaC2oWq+mnAXP2T3XsrHGBejA4bIA6efxaGq3mde7AoJJRwzXIsgtFsTLN+hh6UwDN1TjhmuuhdE+gIFplBp00FRwB/3wCPhgHM7JgdNE6bgGGI3AVLnGwJTbWrU4LCKwQhIaDJOrlHBAoeeaviwojjLJdJWPQVa7KFaaYRY9N5G5+f5vwkWlBKa+DjkXbOpvMsrzpWnSI2BjVR2QXgCT7zYEo+FBKCk2NPWUWZqyub2l8Xdd1o4B6PvJKNNzqnUYCSIB8RErndRZQRske0u5KtfeaFpzGgZnwjRWKnTvNkiyz1mVxxSxWimze02x4FpMc0k5Pyg2f8TXfh9jnH049Oi2yNkJC7WjPWV49vmGe7qfwMYsWkXsK+AcvVFt9JhpT61ekO0LXnL7kn5QqVdffrYPqP2xqWg0MvvE54z+w+crrZ/tjstxyqMNs/4Xt+MRjg==).
